File: native/app/group_events.go
package app

import (
	"bytes"
	"errors"
	"time"

	"github.com/JohnnyPBelo/relayloom/native/core"
	"github.com/JohnnyPBelo/relayloom/native/groupaccess"
	"github.com/JohnnyPBelo/relayloom/native/groupauthority"
	"github.com/JohnnyPBelo/relayloom/native/groupledger"
	"github.com/JohnnyPBelo/relayloom/native/groupstore"
	"github.com/JohnnyPBelo/relayloom/native/profilestate"
	"github.com/JohnnyPBelo/relayloom/native/transport"
)

func currentGroupEvent(kind string) bool {
	return kind == "edit" || kind == "reaction" || kind == "comment"
}

// Original verified bytes plus a durable local admission determine the target;
// caller-provided summaries, keys and contacts cannot grant signing authority.
func (n *Node) commitGroupEventLocked(input Content, recipients any, ttl int64) (core.Bundle, error) {
	var bundle core.Bundle
	if ttl != 0 && (ttl < 1000 || ttl > core.MaxTTL) {
		return bundle, errors.New("prazo inválido")
	}
	if err := validateContent(input); err != nil {
		return bundle, err
	}
	binding, err := groupaccess.ParseBinding(input)
	if err != nil {
		return bundle, err
	}
	kind := text(input["type"])
	if (!currentGroupEvent(kind) && kind != "delete") || text(recipients) == "public" {
		return bundle, errors.New("evento privado de grupo inválido")
	}
	requested, err := stringsList(recipients, 64, true)
	if err != nil {
		return bundle, err
	}
	requested = append(requested, n.identity.Public.ID)
	historical := kind == "delete"
	expectedAudience := "target"
	if historical {
		expectedAudience = "historical"
	}
	if binding.Audience != expectedAudience {
		return bundle, errors.New("audiência de evento inválida")
	}
	fields := []string{"type", "target", "conversation", "targetEpoch", "groupAudience"}
	if !historical {
		fields = append(fields, "groupEpoch")
		if kind == "reaction" {
			fields = append(fields, "emoji", "value")
		} else {
			fields = append(fields, "text")
		}
	}
	for field := range input {
		if !contains(fields, field) {
			return bundle, errors.New("campos do evento de grupo inválidos")
		}
	}
	if !historical && kind != "reaction" && trimmed(text(input["text"])) == "" {
		return bundle, errors.New("texto do evento inválido")
	}
	if kind == "reaction" {
		if _, ok := input["value"].(bool); !ok || trimmed(text(input["emoji"])) == "" {
			return bundle, errors.New("reacção inválida")
		}
	}
	original, err := n.Store.GetWithTouch(text(input["target"]), false)
	if err != nil {
		return bundle, err
	}
	targetObject, err := n.displayLocked(original)
	if err != nil {
		return bundle, err
	}
	targetBinding, err := groupaccess.ParseBinding(targetObject.Content)
	if err != nil {
		return bundle, err
	}
	if targetObject.Public || targetObject.Kind != "message" || targetBinding.Conversation != binding.Conversation || targetBinding.Epoch != binding.TargetEpoch || !contains(targetObject.Readers, n.identity.Public.ID) || ((historical || kind == "edit") && targetObject.Author.ID != n.identity.Public.ID) {
		return bundle, errors.New("sem autoridade para alterar este alvo de grupo")
	}
	protected := n.protectedGroupContentLocked()
	update := n.updateGroupState
	if update == nil {
		update = n.privateDatabase.Update
	}
	err = update(func(tx *groupstore.Tx) error {
		current, err := profilestate.Read(tx)
		if err != nil {
			return err
		}
		if current == nil || current.Digest != n.privateDigest {
			return errors.New("o estado privado mudou antes do evento")
		}
		_, err = groupledger.Run(tx, *n.identity, func(l *groupledger.Ledger, g *groupauthority.Registry) error {
			target, err := l.Accepted(original.Manifest.ID)
			if err != nil {
				return err
			}
			if target == nil || target.Context.Kind != "message" || target.Context.Author != targetObject.Author.ID || target.Context.GroupID != binding.Conversation || target.Context.EpochID != binding.TargetEpoch || target.Expires != targetObject.Expires || !equalIDs(target.Context.Readers, targetObject.Readers) {
				return errors.New("o evento exige um alvo localmente admitido")
			}
			epoch := binding.Epoch
			if historical {
				epoch = binding.TargetEpoch
			} else {
				state, err := g.State(binding.Conversation)
				if err != nil {
					return err
				}
				if state.Status != "active" || state.Head == nil || state.Head.ID != epoch {
					return errors.New("o grupo mudou; reveja a audiência antes de publicar")
				}
			}
			snapshot, err := g.PrivateState(binding.Conversation, epoch)
			if err != nil {
				return err
			}
			if snapshot == nil {
				return errors.New("falta confirmar o estado privado do grupo")
			}
			cards := []core.PublicIdentity{}
			for _, card := range snapshot.Members {
				if contains(targetObject.Readers, card.ID) {
					cards = append(cards, card)
				}
			}
			ids := memberIDs(cards)
			unique := []string{}
			for _, id := range requested {
				if !contains(unique, id) {
					unique = append(unique, id)
				}
			}
			if !contains(ids, n.identity.Public.ID) || !equalIDs(ids, unique) {
				return errors.New("destinatários diferentes da audiência autorizada")
			}
			for _, id := range ids {
				if contains(n.config.Blocked, id) {
					return errors.New("há um destinatário bloqueado")
				}
			}
			remaining := targetObject.Expires - time.Now().UnixMilli()
			if remaining <= 0 {
				return errors.New("o conteúdo original expirou")
			}
			if ttl == 0 {
				ttl = remaining
			}
			bundle, err = core.CreateBundle(*n.identity, kind, input, cards, false, max(1000, min(ttl, remaining)))
			if err != nil {
				return err
			}
			if err = core.VerifyBundle(bundle); err != nil {
				return err
			}
			decoded, err := core.DecryptBundle(bundle, n.identity)
			if err != nil {
				return err
			}
			verified, err := object(decoded)
			if err != nil {
				return err
			}
			if err = validateContent(Content(verified)); err != nil {
				return err
			}
			a, err := core.Canonical(input)
			if err != nil {
				return err
			}
			b, err := core.Canonical(verified)
			if err != nil {
				return err
			}
			if !bytes.Equal(a, b) {
				return errors.New("evento assinado incoerente")
			}
			encoded, err := core.Canonical(bundle)
			if err != nil {
				return err
			}
			result, err := l.Consider(groupaccess.Candidate{ID: bundle.Manifest.ID, Kind: kind, Author: n.identity.Public, Readers: ids, Content: verified}, bundle.Manifest.Expires, int64(len(encoded)), protected, time.Now().UnixMilli())
			if err != nil {
				return err
			}
			if result.Decision.Status != "accepted" {
				return errors.New("autoridade do evento não confirmada")
			}
			return nil
		})
		return err
	})
	if err != nil {
		n.recoverGroupContentLocked()
	}
	return bundle, err
}

func (n *Node) publishGroupEventLocked(content Content, recipients any, ttl int64) (DisplayObject, error) {
	if n.identity == nil || n.privateDatabase == nil {
		return DisplayObject{}, errors.New("desbloqueie a identidade")
	}
	if _, err := n.objectsLocked(); err != nil {
		return DisplayObject{}, err
	}
	bundle, err := n.commitGroupEventLocked(content, recipients, ttl)
	if err != nil {
		return DisplayObject{}, err
	}
	if _, err = n.Store.Put(bundle, false); err != nil {
		return DisplayObject{}, err
	}
	accepted, err := n.observeGroupLocked(bundle.Manifest.ID, n.protectedGroupContentLocked())
	if err != nil {
		return DisplayObject{}, err
	}
	if !accepted {
		return DisplayObject{}, errors.New("o evento não pôde ser admitido")
	}
	if err = n.reconcileGroupSendsLocked(); err != nil {
		return DisplayObject{}, err
	}
	allowed, err := n.maySeedLocked(bundle.Manifest)
	if err != nil {
		return DisplayObject{}, err
	}
	if !allowed {
		return DisplayObject{}, errors.New("o grupo já não autoriza este evento")
	}
	if _, err = n.Router.Broadcast(map[string]any{"type": "bundle", "bundle": bundle}, transport.Normal, 2*time.Minute, false); err != nil {
		return DisplayObject{}, err
	}
	result, err := n.displayLocked(bundle)
	if err != nil {
		return DisplayObject{}, err
	}
	return *result, nil
}

func (n *Node) localGroupEventsLocked(manifests map[string]core.Manifest) []DisplayObject {
	events := []DisplayObject{}
	for _, manifest := range manifests {
		if manifest.Author.ID != n.identity.Public.ID || manifest.PublicKey != nil || !currentGroupEvent(manifest.Kind) {
			continue
		}
		bundle, err := n.Store.GetWithTouch(manifest.ID, false)
		if err != nil {
			continue
		}
		value, err := n.displayLocked(bundle)
		if err != nil || !groupaccess.HasBinding(value.Content) {
			continue
		}
		if _, err = groupaccess.ParseBinding(value.Content); err != nil {
			continue
		}
		events = append(events, *value)
	}
	return events
}

func eventSeeds(l *groupledger.Ledger, events []DisplayObject, blocked []string, access *groupaccess.Access) (map[string]bool, error) {
	allowed := map[string]bool{}
	for _, event := range events {
		accepted, err := l.Accepted(event.ID)
		if err != nil {
			return nil, err
		}
		if accepted == nil || accepted.Context.Kind != event.Kind || accepted.Context.Author != event.Author.ID || accepted.Context.GroupID != text(event.Content["conversation"]) || accepted.Context.EpochID != text(event.Content["groupEpoch"]) || accepted.Expires != event.Expires || !equalIDs(accepted.Context.Readers, event.Readers) {
			continue
		}
		deny := false
		for _, id := range event.Readers {
			if contains(blocked, id) {
				deny = true
			}
		}
		if deny {
			continue
		}
		target, err := l.Accepted(text(event.Content["target"]))
		if err != nil {
			return nil, err
		}
		var targetContext *groupaccess.AcceptedContext
		if target != nil {
			targetContext = &target.Context
		}
		decision, err := access.Decide(groupaccess.Candidate{ID: event.ID, Kind: event.Kind, Author: event.Author, Readers: event.Readers, Content: event.Content}, &accepted.Context, targetContext)
		if err != nil {
			return nil, err
		}
		if decision.Status != "accepted" {
			continue
		}
		authority, err := l.RetryAuthority(accepted.Context.GroupID, accepted.Context.EpochID)
		if err != nil {
			return nil, err
		}
		if authority.Allowed {
			allowed[event.ID] = true
		}
	}
	return allowed, nil
}

File: apps/node/src/group-events.ts
import {
  canonical,
  createBundle,
  decryptBundle,
  verifyBundle,
  type Identity,
  type ContentStore,
  type Bundle,
  type Manifest,
} from "../../../packages/core/src/index.js";
import type { ProfileDatabase } from "../../../packages/profile/src/database.js";
import { readProfileState } from "../../../packages/profile/src/state.js";
import { GroupLedger } from "../../../packages/groups/src/ledger.js";
import {
  GroupAccess,
  hasGroupBinding,
  parseGroupBinding,
} from "../../../packages/groups/src/access.js";
import { validateContent } from "./content-validation.js";
import type { Content } from "./node.js";

export const currentGroupEvent = (kind: string) =>
  ["edit", "reaction", "comment"].includes(kind);

/** The original bytes and authenticated admission establish the target. The
 * caller cannot supply a historical context, reader key or replacement card. */
export function commitGroupEvent(
  database: ProfileDatabase,
  identity: Identity,
  privateDigest: string,
  store: ContentStore,
  input: Content,
  recipients: string[] | "public",
  blocked: readonly string[],
  ttlMs?: number,
) {
  validateContent(input);
  if (
    ttlMs !== undefined &&
    (!Number.isSafeInteger(ttlMs) || ttlMs < 1000 || ttlMs > 365 * 86400_000)
  )
    throw new Error("Prazo inválido");
  if (
    recipients !== "public" &&
    (!Array.isArray(recipients) ||
      recipients.length > 64 ||
      recipients.some(
        (id) => typeof id !== "string" || !/^[a-f0-9]{64}$/.test(id),
      ))
  )
    throw new Error("Destinatários inválidos");
  const binding = parseGroupBinding(input);
  if (
    (!currentGroupEvent(input.type) && input.type !== "delete") ||
    recipients === "public"
  )
    throw new Error("Evento privado de grupo inválido");
  const historical = input.type === "delete";
  if (binding.groupAudience !== (historical ? "historical" : "target"))
    throw new Error("Audiência de evento inválida");
  const allowedFields = [
    "type",
    "target",
    "conversation",
    "targetEpoch",
    "groupAudience",
    ...(historical
      ? []
      : [
          "groupEpoch",
          ...(input.type === "reaction" ? ["emoji", "value"] : ["text"]),
        ]),
  ];
  if (
    Object.keys(input).some((key) => !allowedFields.includes(key)) ||
    (!historical && input.type !== "reaction" && !input.text?.trim()) ||
    (input.type === "reaction" &&
      (!input.emoji?.trim() || typeof input.value !== "boolean"))
  )
    throw new Error("Campos do evento de grupo inválidos");
  const original = store.get(input.target!, false),
    targetContent = decryptBundle(original, identity) as Content;
  validateContent(targetContent);
  const targetBinding = parseGroupBinding(targetContent),
    originalReaders = original.manifest.keys.map((key) => key.reader).sort();
  if (
    original.manifest.publicKey ||
    original.manifest.kind !== "message" ||
    targetContent.type !== "message" ||
    targetBinding.conversation !== binding.conversation ||
    targetBinding.groupEpoch !== binding.targetEpoch ||
    !originalReaders.includes(identity.public.id) ||
    ((historical || input.type === "edit") &&
      original.manifest.author.id !== identity.public.id)
  )
    throw new Error("Sem autoridade para alterar este alvo de grupo");
  const protectedIds = new Set(store.list().map((manifest) => manifest.id));
  return database.transaction((tx) => {
    if (readProfileState(tx)?.digest !== privateDigest)
      throw new Error("O estado privado mudou antes do evento");
    return GroupLedger.run(tx, identity, (ledger, registry) => {
      const target = ledger.accepted(original.manifest.id);
      if (
        !target ||
        target.context.kind !== "message" ||
        target.context.author !== original.manifest.author.id ||
        target.context.groupId !== binding.conversation ||
        target.context.epochId !== binding.targetEpoch ||
        target.expires !== original.manifest.expires ||
        canonical(target.context.readers) !== canonical(originalReaders)
      )
        throw new Error("O evento exige um alvo localmente admitido");
      const epoch = historical ? binding.targetEpoch! : binding.groupEpoch!;
      if (!historical) {
        const state = registry.state(binding.conversation);
        if (state.status !== "active" || state.head?.id !== epoch)
          throw new Error(
            "O grupo mudou. Reveja a audiência antes de publicar.",
          );
      }
      const snapshot = registry.privateState(binding.conversation, epoch);
      if (!snapshot)
        throw new Error("Falta confirmar o estado privado do grupo");
      const cards = snapshot.members.filter((card) =>
        originalReaders.includes(card.id),
      );
      const readers = cards.map((card) => card.id).sort();
      if (
        !readers.includes(identity.public.id) ||
        readers.some((id) => blocked.includes(id)) ||
        canonical(readers) !==
          canonical([...new Set([...recipients, identity.public.id])].sort())
      )
        throw new Error("Destinatários diferentes da audiência autorizada");
      const remaining = original.manifest.expires - Date.now();
      if (remaining <= 0) throw new Error("O conteúdo original expirou");
      const bundle = createBundle(
        identity,
        input.type,
        structuredClone(input),
        cards,
        Math.max(1000, Math.min(ttlMs ?? remaining, remaining)),
      );
      verifyBundle(bundle);
      const verified = decryptBundle(bundle, identity) as Content;
      validateContent(verified);
      if (canonical(verified) !== canonical(input))
        throw new Error("Evento assinado incoerente");
      const admission = ledger.consider(
        {
          id: bundle.manifest.id,
          kind: input.type,
          author: identity.public,
          readers,
          public: false,
          content: verified,
        },
        bundle.manifest.expires,
        Buffer.byteLength(canonical(bundle)),
        protectedIds,
      );
      if (admission.decision.status !== "accepted")
        throw new Error("Autoridade do evento não confirmada");
      return bundle;
    }).value;
  });
}

export function localGroupEvents(
  store: ContentStore,
  identity: Identity,
  manifests: readonly Manifest[] = store.list(),
) {
  const events: { bundle: Bundle; content: Content }[] = [];
  for (const manifest of manifests) {
    if (
      manifest.author.id !== identity.public.id ||
      manifest.publicKey ||
      !currentGroupEvent(manifest.kind)
    )
      continue;
    try {
      const bundle = store.get(manifest.id, false),
        content = decryptBundle(bundle, identity) as Content;
      validateContent(content);
      if (content.type === manifest.kind && hasGroupBinding(content)) {
        parseGroupBinding(content);
        events.push({ bundle, content });
      }
    } catch {
      /* Missing/corrupt bytes cannot grant permission to a queued copy. */
    }
  }
  return events;
}

export function groupEventSeeds(
  ledger: GroupLedger,
  events: ReturnType<typeof localGroupEvents>,
  blocked: readonly string[],
  access: GroupAccess,
) {
  const allowed = new Set<string>();
  for (const { bundle, content } of events) {
    const manifest = bundle.manifest,
      accepted = ledger.accepted(manifest.id);
    const readers = manifest.keys.map((key) => key.reader).sort();
    if (
      !accepted ||
      accepted.context.kind !== manifest.kind ||
      accepted.context.author !== manifest.author.id ||
      accepted.context.groupId !== content.conversation ||
      accepted.context.epochId !== content.groupEpoch ||
      accepted.expires !== manifest.expires ||
      canonical(accepted.context.readers) !== canonical(readers) ||
      readers.some((id) => blocked.includes(id))
    )
      continue;
    const target = ledger.accepted(content.target!);
    const decision = access.decide(
      {
        id: manifest.id,
        kind: manifest.kind,
        author: manifest.author,
        readers,
        public: false,
        content,
      },
      accepted.context,
      target?.context,
    );
    if (
      decision.status === "accepted" &&
      ledger.retryAuthority(content.conversation!, content.groupEpoch!).allowed
    )
      allowed.add(manifest.id);
  }
  return allowed;
}
